File: cartera/predictivo.py
"""Agregaciones predictivas sobre el universo completo de socios en caché."""

from collections import Counter

UMBRAL_ALERTA_PROBABILIDAD = 40.0

BANDAS_SCORE = [
    {"key": "0-25", "label": "0–25%", "min": 0, "max": 25, "fill": "#34d399"},
    {"key": "25-40", "label": "25–40%", "min": 25, "max": 40, "fill": "#3DAA4A"},
    {"key": "40-65", "label": "40–65%", "min": 40, "max": 65, "fill": "#F39237"},
    {"key": "65-100", "label": "65–100%", "min": 65, "max": 100.01, "fill": "#ef4444"},
]

BANDAS_DIAS_MORA = [
    {"key": "0", "label": "Al día (0 d)", "min": 0, "max": 1, "fill": "#3DAA4A"},
    {"key": "1-30", "label": "1–30 días", "min": 1, "max": 31, "fill": "#F39237"},
    {"key": "31-90", "label": "31–90 días", "min": 31, "max": 91, "fill": "#ea580c"},
    {"key": "90+", "label": "Más de 90 d", "min": 91, "max": float("inf"), "fill": "#ef4444"},
]

BANDAS_MONTO = [
    {"key": "bajo", "label": "< $5.000", "min": 0, "max": 5000, "fill": "#3DAA4A"},
    {"key": "medio", "label": "$5k – $15k", "min": 5000, "max": 15000, "fill": "#F39237"},
    {"key": "alto", "label": "$15k – $50k", "min": 15000, "max": 50000, "fill": "#ea580c"},
    {"key": "muy-alto", "label": "> $50.000", "min": 50000, "max": float("inf"), "fill": "#ef4444"},
]

BANDAS_AHORRO = [
    {"key": "critico", "label": "< 5% crédito", "min": 0, "max": 0.05, "fill": "#ef4444"},
    {"key": "bajo", "label": "5% – 20%", "min": 0.05, "max": 0.2, "fill": "#F39237"},
    {"key": "medio", "label": "20% – 50%", "min": 0.2, "max": 0.5, "fill": "#fbbf24"},
    {"key": "saludable", "label": "> 50%", "min": 0.5, "max": float("inf"), "fill": "#3DAA4A"},
]

COLORES_CALIF = ["#3DAA4A", "#1F6B3D", "#F39237", "#ea580c", "#ef4444", "#6366f1", "#8b5cf6", "#0ea5e9"]
COLORES_TIPO = ["#3DAA4A", "#F39237", "#1F6B3D", "#ef4444", "#0ea5e9", "#a855f7", "#64748b"]


def _variante_riesgo(prob: float) -> str:
    if prob >= 65:
        return "critico"
    if prob >= 40:
        return "moderado"
    return "estable"


def _en_banda(valor: float, bandas: list) -> str | None:
    for banda in bandas:
        if valor >= banda["min"] and valor < banda["max"]:
            return banda["key"]
    return None


def _pasa_filtros(socio: dict, filtros: dict) -> bool:
    prob = socio["probabilidad_mora"]
    variant = _variante_riesgo(prob)

    tipo = filtros.get("tipo_riesgo") or "todos"
    if tipo != "todos" and variant != tipo:
        return False

    rango = filtros.get("rango_score") or "todos"
    if rango == "alto" and prob < 65:
        return False
    if rango == "medio" and (prob < 40 or prob >= 65):
        return False
    if rango == "bajo" and prob >= 40:
        return False

    mora = filtros.get("solo_mora") or "todos"
    if mora == "si" and socio["dias_mora"] <= 0:
        return False
    if mora == "no" and socio["dias_mora"] > 0:
        return False

    calif = filtros.get("calificacion") or "todas"
    if calif != "todas":
        if (socio.get("calificacion") or "").strip().upper() != calif.upper():
            return False

    return True


def _histograma_desde_conteo(bandas: list, conteo: dict, total: int) -> list:
    return [
        {
            "key": b["key"],
            "label": b["label"],
            "fill": b["fill"],
            "cantidad": conteo.get(b["key"], 0),
            "porcentaje": round((conteo.get(b["key"], 0) / total) * 100) if total else 0,
        }
        for b in bandas
    ]


def _pastel_desde_conteo(items: list[tuple], colores: list, limite: int = 8) -> list:
    ordenados = sorted(items, key=lambda x: x[1], reverse=True)
    top = ordenados[:limite]
    otros = sum(v for _, v in ordenados[limite:])
    segmentos = []
    for i, (label, valor) in enumerate(top):
        if valor <= 0:
            continue
        segmentos.append(
            {
                "id": str(label),
                "label": str(label)[:20],
                "labelCompleto": str(label),
                "valor": valor,
                "color": colores[i % len(colores)],
            }
        )
    if otros > 0:
        segmentos.append(
            {"id": "otros", "label": "Otros", "valor": otros, "color": "#94a3b8"}
        )
    total = sum(s["valor"] for s in segmentos) or 1
    for s in segmentos:
        s["porcentaje"] = round((s["valor"] / total) * 100)
    return segmentos


def calcular_resumen_predictivo(credito: dict, ahorro: dict, filtros: dict) -> dict:
    from .views import _codigo_oficina, _construir_socio, _filtrar_tokens_por_oficina

    oficina = (filtros.get("oficina") or "").strip()
    tokens = sorted(credito.keys())
    if oficina:
        tokens = _filtrar_tokens_por_oficina(tokens, credito, oficina)

    califs_universo = set()
    for token in tokens:
        cal = (credito[token].get("calificacion") or "").strip().upper()
        if cal:
            califs_universo.add(cal)

    conteo_score = {b["key"]: 0 for b in BANDAS_SCORE}
    conteo_mora_dias = {b["key"]: 0 for b in BANDAS_DIAS_MORA}
    conteo_monto = {b["key"]: 0 for b in BANDAS_MONTO}
    conteo_ahorro = {b["key"]: 0 for b in BANDAS_AHORRO}

    criticos = moderados = estables = 0
    con_mora = sin_mora = 0
    capital_critico = capital_moderado = capital_estable = 0.0
    suma_score = 0.0
    alertas = 0
    capital_riesgo = 0.0

    conteo_calif = Counter()
    conteo_tipo = Counter()
    conteo_agencia = {}

    top_candidatos: list[dict] = []

    total_filtrado = 0

    for token in tokens:
        fila_c = credito[token]
        fila_a = ahorro.get(token, {})
        socio = _construir_socio(token, fila_c, fila_a)

        if not _pasa_filtros(socio, filtros):
            continue

        total_filtrado += 1
        prob = socio["probabilidad_mora"]
        dias = socio["dias_mora"]
        monto = socio["monto_credito"]
        ahorro_val = socio["saldo_disponible"]

        suma_score += prob
        if prob >= UMBRAL_ALERTA_PROBABILIDAD or dias > 0:
            alertas += 1
        if prob >= 40:
            capital_riesgo += monto

        variant = _variante_riesgo(prob)
        if variant == "critico":
            criticos += 1
            capital_critico += monto
        elif variant == "moderado":
            moderados += 1
            capital_moderado += monto
        else:
            estables += 1
            capital_estable += monto

        if dias > 0:
            con_mora += 1
        else:
            sin_mora += 1

        k_score = _en_banda(prob, BANDAS_SCORE)
        if k_score:
            conteo_score[k_score] += 1
        k_mora = _en_banda(float(dias), BANDAS_DIAS_MORA)
        if k_mora:
            conteo_mora_dias[k_mora] += 1
        k_monto = _en_banda(monto, BANDAS_MONTO)
        if k_monto:
            conteo_monto[k_monto] += 1
        ratio = (ahorro_val / monto) if monto > 0 else 1.0
        k_ahorro = _en_banda(ratio, BANDAS_AHORRO)
        if k_ahorro:
            conteo_ahorro[k_ahorro] += 1

        cal = (socio.get("calificacion") or "Sin cal.").strip().upper() or "SIN CAL."
        conteo_calif[cal] += 1
        tipo = (socio.get("tipo_cartera") or "Sin tipo").strip() or "Sin tipo"
        conteo_tipo[tipo] += 1

        cod_ag = _codigo_oficina(socio.get("nro_oficina")) or "—"
        if cod_ag not in conteo_agencia:
            conteo_agencia[cod_ag] = {"codigo": cod_ag, "total": 0, "alertas": 0, "suma_score": 0.0}
        ag = conteo_agencia[cod_ag]
        ag["total"] += 1
        ag["suma_score"] += prob
        if prob >= 40 or dias > 0:
            ag["alertas"] += 1

        if len(top_candidatos) < 10:
            top_candidatos.append(socio)
            top_candidatos.sort(key=lambda s: (-s["probabilidad_mora"], -s["dias_mora"]))
        else:
            peor = top_candidatos[-1]
            if (prob, dias) > (peor["probabilidad_mora"], peor["dias_mora"]):
                top_candidatos[-1] = socio
                top_candidatos.sort(key=lambda s: (-s["probabilidad_mora"], -s["dias_mora"]))

    total = total_filtrado or 1
    tasa_alerta = (alertas / total_filtrado * 100) if total_filtrado else 0.0
    tasa_mora = (con_mora / total_filtrado * 100) if total_filtrado else 0.0

    por_agencia = sorted(
        [
            {
                **ag,
                "tasaAlerta": round((ag["alertas"] / ag["total"]) * 100, 1) if ag["total"] else 0,
                "scorePromedio": round(ag["suma_score"] / ag["total"], 1) if ag["total"] else 0,
            }
            for ag in conteo_agencia.values()
        ],
        key=lambda x: x["tasaAlerta"],
        reverse=True,
    )[:12]

    pastel_riesgo = _pastel_desde_conteo(
        [("Crítico", criticos), ("Moderado", moderados), ("Estable", estables)],
        ["#ef4444", "#F39237", "#3DAA4A"],
        limite=3,
    )
    pastel_mora = _pastel_desde_conteo(
        [("Con mora", con_mora), ("Al día", sin_mora)],
        ["#ef4444", "#3DAA4A"],
        limite=2,
    )

    return {
        "universo_completo": True,
        "total_universo_oficina": len(tokens),
        "total_analizados": total_filtrado,
        "metricas": {
            "score_promedio": round(suma_score / total_filtrado, 2) if total_filtrado else 0,
            "tasa_alerta": round(tasa_alerta, 2),
            "tasa_mora_real": round(tasa_mora, 2),
            "capital_en_riesgo": round(capital_riesgo, 2),
            "criticos": criticos,
            "moderados": moderados,
            "estables": estables,
            "alertas": alertas,
        },
        "histogramas": {
            "score": _histograma_desde_conteo(BANDAS_SCORE, conteo_score, total_filtrado),
            "dias_mora": _histograma_desde_conteo(BANDAS_DIAS_MORA, conteo_mora_dias, total_filtrado),
            "montos": _histograma_desde_conteo(BANDAS_MONTO, conteo_monto, total_filtrado),
            "ahorro_ratio": _histograma_desde_conteo(BANDAS_AHORRO, conteo_ahorro, total_filtrado),
        },
        "pasteles": {
            "riesgo": pastel_riesgo,
            "mora": pastel_mora,
            "calificacion": _pastel_desde_conteo(conteo_calif.items(), COLORES_CALIF, 8),
            "tipo_cartera": _pastel_desde_conteo(conteo_tipo.items(), COLORES_TIPO, 6),
            "capital_riesgo": _pastel_desde_conteo(
                [
                    ("Capital crítico", int(capital_critico)),
                    ("Capital moderado", int(capital_moderado)),
                    ("Capital estable", int(capital_estable)),
                ],
                ["#ef4444", "#F39237", "#3DAA4A"],
                3,
            ),
        },
        "por_agencia": por_agencia,
        "top_riesgo": sorted(
            top_candidatos,
            key=lambda s: (-s["probabilidad_mora"], -s["dias_mora"]),
        ),
        "calificaciones_disponibles": sorted(califs_universo),
    }

File: frontend/src/utils/predictivo.js
import { nivelRiesgo } from "./risk.js";

export const BANDAS_HISTOGRAMA = [
  { key: "0-25", label: "0–25%", min: 0, max: 25, color: "bg-emerald-400" },
  { key: "25-40", label: "25–40%", min: 25, max: 40, color: "bg-coop-green" },
  { key: "40-65", label: "40–65%", min: 40, max: 65, color: "bg-coop-orange" },
  { key: "65-100", label: "65–100%", min: 65, max: 100.01, color: "bg-red-500" },
];

export function filtrarSociosPredictivo(socios, filtros) {
  const { tipoRiesgo, rangoScore, soloMora, calificacion } = filtros;

  return socios.filter((socio) => {
    const score = socio.probabilidad_mora ?? 0;
    const riesgo = nivelRiesgo(score).variant;

    if (tipoRiesgo !== "todos" && riesgo !== tipoRiesgo) return false;
    if (rangoScore === "alto" && score < 65) return false;
    if (rangoScore === "medio" && (score < 40 || score >= 65)) return false;
    if (rangoScore === "bajo" && score >= 40) return false;
    if (soloMora === "si" && !(socio.dias_mora > 0)) return false;
    if (soloMora === "no" && socio.dias_mora > 0) return false;
    if (calificacion !== "todas") {
      const cal = (socio.calificacion || "").trim().toUpperCase();
      if (cal !== calificacion.toUpperCase()) return false;
    }
    return true;
  });
}

export function histogramaScores(socios) {
  const conteo = Object.fromEntries(BANDAS_HISTOGRAMA.map((b) => [b.key, 0]));
  for (const socio of socios) {
    const score = socio.probabilidad_mora ?? 0;
    const banda = BANDAS_HISTOGRAMA.find((b) => score >= b.min && score < b.max);
    if (banda) conteo[banda.key] += 1;
  }
  const total = socios.length || 1;
  return BANDAS_HISTOGRAMA.map((b) => ({
    ...b,
    cantidad: conteo[b.key],
    porcentaje: Math.round((conteo[b.key] / total) * 100),
  }));
}

export function metricasPredictivo(socios, escenarioStressPct = 15) {
  const n = socios.length;
  if (!n) {
    return {
      scorePromedio: 0,
      tasaAlerta: 0,
      tasaMoraReal: 0,
      capitalEnRiesgo: 0,
      proyeccionStress: 0,
      criticos: 0,
      moderados: 0,
      estables: 0,
    };
  }

  let sumaScore = 0;
  let alertas = 0;
  let conMora = 0;
  let capitalRiesgo = 0;
  let criticos = 0;
  let moderados = 0;
  let estables = 0;

  for (const socio of socios) {
    const score = socio.probabilidad_mora ?? 0;
    sumaScore += score;
    if (score >= 40 || socio.dias_mora > 0) alertas += 1;
    if (socio.dias_mora > 0) conMora += 1;
    if (score >= 40) capitalRiesgo += socio.monto_credito || 0;

    const v = nivelRiesgo(score).variant;
    if (v === "critico") criticos += 1;
    else if (v === "moderado") moderados += 1;
    else estables += 1;
  }

  const tasaAlerta = (alertas / n) * 100;
  const factor = 1 + escenarioStressPct / 100;

  return {
    scorePromedio: sumaScore / n,
    tasaAlerta,
    tasaMoraReal: (conMora / n) * 100,
    capitalEnRiesgo: capitalRiesgo,
    proyeccionStress: Math.min(tasaAlerta * factor, 100),
    criticos,
    moderados,
    estables,
  };
}

export function concentracionPorAgencia(socios, limite = 8) {
  const mapa = new Map();

  for (const socio of socios) {
    const codigo = String(socio.nro_oficina ?? socio.credito?.nro_oficina ?? "—");
    if (!mapa.has(codigo)) {
      mapa.set(codigo, { codigo, total: 0, alertas: 0, sumaScore: 0 });
    }
    const item = mapa.get(codigo);
    item.total += 1;
    item.sumaScore += socio.probabilidad_mora ?? 0;
    if ((socio.probabilidad_mora ?? 0) >= 40 || socio.dias_mora > 0) {
      item.alertas += 1;
    }
  }

  return [...mapa.values()]
    .map((item) => ({
      ...item,
      tasaAlerta: item.total ? (item.alertas / item.total) * 100 : 0,
      scorePromedio: item.total ? item.sumaScore / item.total : 0,
    }))
    .sort((a, b) => b.tasaAlerta - a.tasaAlerta)
    .slice(0, limite);
}

export function topSociosRiesgo(socios, limite = 10) {
  return [...socios]
    .sort((a, b) => {
      const diff = (b.probabilidad_mora ?? 0) - (a.probabilidad_mora ?? 0);
      if (diff !== 0) return diff;
      return (b.dias_mora ?? 0) - (a.dias_mora ?? 0);
    })
    .slice(0, limite);
}

export function calificacionesDisponibles(socios) {
  const set = new Set();
  for (const s of socios) {
    const cal = (s.calificacion || "").trim();
    if (cal) set.add(cal.toUpperCase());
  }
  return [...set].sort();
}

export const BANDAS_DIAS_MORA = [
  { key: "0", label: "Al día (0 d)", min: 0, max: 1, color: "bg-coop-green", fill: "#3DAA4A" },
  { key: "1-30", label: "1–30 días", min: 1, max: 31, color: "bg-amber-400", fill: "#F39237" },
  { key: "31-90", label: "31–90 días", min: 31, max: 91, color: "bg-orange-500", fill: "#ea580c" },
  { key: "90+", label: "Más de 90 d", min: 91, max: Infinity, color: "bg-red-500", fill: "#ef4444" },
];

export const BANDAS_MONTO_CREDITO = [
  { key: "bajo", label: "< $5.000", min: 0, max: 5000, color: "bg-coop-green", fill: "#3DAA4A" },
  { key: "medio", label: "$5k – $15k", min: 5000, max: 15000, color: "bg-coop-orange", fill: "#F39237" },
  { key: "alto", label: "$15k – $50k", min: 15000, max: 50000, color: "bg-orange-500", fill: "#ea580c" },
  { key: "muy-alto", label: "> $50.000", min: 50000, max: Infinity, color: "bg-red-500", fill: "#ef4444" },
];

export const BANDAS_RATIO_AHORRO = [
  { key: "critico", label: "< 5% crédito", min: 0, max: 0.05, color: "bg-red-500", fill: "#ef4444" },
  { key: "bajo", label: "5% – 20%", min: 0.05, max: 0.2, color: "bg-coop-orange", fill: "#F39237" },
  { key: "medio", label: "20% – 50%", min: 0.2, max: 0.5, color: "bg-amber-400", fill: "#fbbf24" },
  { key: "saludable", label: "> 50%", min: 0.5, max: Infinity, color: "bg-coop-green", fill: "#3DAA4A" },
];

function _histogramaDesdeBandas(socios, bandas, obtenerValor) {
  const conteo = Object.fromEntries(bandas.map((b) => [b.key, 0]));
  for (const socio of socios) {
    const valor = obtenerValor(socio);
    const banda = bandas.find((b) => valor >= b.min && valor < b.max);
    if (banda) conteo[banda.key] += 1;
  }
  const total = socios.length || 1;
  return bandas.map((b) => ({
    ...b,
    cantidad: conteo[b.key],
    porcentaje: Math.round((conteo[b.key] / total) * 100),
  }));
}

export function histogramaDiasMora(socios) {
  return _histogramaDesdeBandas(socios, BANDAS_DIAS_MORA, (s) => Math.max(0, s.dias_mora ?? 0));
}

export function histogramaMontosCredito(socios) {
  return _histogramaDesdeBandas(socios, BANDAS_MONTO_CREDITO, (s) => s.monto_credito ?? 0);
}

export function histogramaRatioAhorro(socios) {
  return _histogramaDesdeBandas(socios, BANDAS_RATIO_AHORRO, (s) => {
    const credito = s.monto_credito || 0;
    const ahorro = s.saldo_disponible ?? 0;
    if (credito <= 0) return 1;
    return ahorro / credito;
  });
}

function _aPorcentajes(segmentos) {
  const total = segmentos.reduce((s, x) => s + x.valor, 0) || 1;
  return segmentos.map((s) => ({
    ...s,
    porcentaje: Math.round((s.valor / total) * 100),
  }));
}

export function pastelRiesgo(stats) {
  return _aPorcentajes([
    { id: "critico", label: "Crítico", valor: stats.criticos, color: "#ef4444" },
    { id: "moderado", label: "Moderado", valor: stats.moderados, color: "#F39237" },
    { id: "estable", label: "Estable", valor: stats.estables, color: "#3DAA4A" },
  ]).filter((s) => s.valor > 0);
}

export function pastelMora(socios) {
  const conMora = socios.filter((s) => (s.dias_mora ?? 0) > 0).length;
  const alDia = socios.length - conMora;
  return _aPorcentajes([
    { id: "mora", label: "Con mora", valor: conMora, color: "#ef4444" },
    { id: "aldia", label: "Al día", valor: alDia, color: "#3DAA4A" },
  ]).filter((s) => s.valor > 0);
}

export function pastelCalificacion(socios, limite = 6) {
  const mapa = new Map();
  for (const s of socios) {
    const cal = (s.calificacion || "Sin cal.").trim().toUpperCase() || "SIN CAL.";
    mapa.set(cal, (mapa.get(cal) || 0) + 1);
  }
  const colores = ["#3DAA4A", "#1F6B3D", "#F39237", "#ea580c", "#ef4444", "#6366f1", "#8b5cf6"];
  const items = [...mapa.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, limite)
    .map(([label, valor], i) => ({
      id: label,
      label,
      valor,
      color: colores[i % colores.length],
    }));
  return _aPorcentajes(items);
}

export function pastelTipoCartera(socios, limite = 5) {
  const mapa = new Map();
  for (const s of socios) {
    const tipo = (s.tipo_cartera || "Sin tipo").trim() || "Sin tipo";
    mapa.set(tipo, (mapa.get(tipo) || 0) + 1);
  }
  const colores = ["#3DAA4A", "#F39237", "#1F6B3D", "#ef4444", "#0ea5e9", "#a855f7"];
  const items = [...mapa.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, limite)
    .map(([label, valor], i) => ({
      id: label,
      label: label.length > 18 ? `${label.slice(0, 16)}…` : label,
      labelCompleto: label,
      valor,
      color: colores[i % colores.length],
    }));
  return _aPorcentajes(items);
}

export function pastelCapitalPorRiesgo(socios) {
  let critico = 0;
  let moderado = 0;
  let estable = 0;
  for (const s of socios) {
    const monto = s.monto_credito || 0;
    const v = nivelRiesgo(s.probabilidad_mora ?? 0).variant;
    if (v === "critico") critico += monto;
    else if (v === "moderado") moderado += monto;
    else estable += monto;
  }
  return _aPorcentajes([
    { id: "c", label: "Capital crítico", valor: critico, color: "#ef4444" },
    { id: "m", label: "Capital moderado", valor: moderado, color: "#F39237" },
    { id: "e", label: "Capital estable", valor: estable, color: "#3DAA4A" },
  ]).filter((s) => s.valor > 0);
}
